### PR TITLE
adjust declared types to work with dart 2 typing at runtime

### PR DESCRIPTION
**Pull request by devoncarew**
Fix several issues when running flutter_tools in --preview-dart-2:

- fix declared return types to address strong mode at runtime errors
- `cast()` data structures parsed from json so that the types at runtime match what we expect
- add generic types to some methods to allow the types to flow through a method call

We'll need to change the flutter_tools tests over to running in `--preview-dart-2` in order to keep the code dart 2 safe, but we're not quite ready to do that (our test suite is reparsed for each individual test - the test suite takes 40-50 minutes to run).

### CODE REVIEW

**Comment by devoncarew** [comment]
Follow up to https://github.com/flutter/flutter/pull/18960 and #18998.

**Comment by devoncarew** [comment]
We'll also want to update to the (just recently published) latest version of `json_schema`.

File: packages/flutter_tools/lib/src/asset.dart [hunk 1]
```diff
@@ -268,20 +269,21 @@ Map<String, dynamic> _readMaterialFontsManifest() {
...
-  return loadYaml(fs.file(fontsPath).readAsStringSync());
+  return castStringKeyedMap(loadYaml(fs.file(fontsPath).readAsStringSync()));
 }
 
 final Map<String, dynamic> _materialFontsManifest = _readMaterialFontsManifest();
 
 List<Map<String, dynamic>> _getMaterialFonts(String fontSet) {
-  return _materialFontsManifest[fontSet];
+  final List<dynamic> fontsList = _materialFontsManifest[fontSet];
+  return fontsList?.map(castStringKeyedMap)?.toList();
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/asset.dart]: `List.map` should be given a type parameter per [Flutter style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-var):
```dart
return fontsList?.map<Map<String, dynamic>>(castStringKeyedMap)?.toList();
```

File: packages/flutter_tools/lib/src/commands/daemon.dart [hunk 2]
```diff
@@ -184,14 +188,18 @@ abstract class Domain {
       if (_handlers.containsKey(command))
         return _handlers[command](args);
       throw 'command not understood: $name.$command';
-    }).then<Null>((dynamic result) {
+    }).then<dynamic>((dynamic result) {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/commands/daemon.dart]: `then<void>`?

File: packages/flutter_tools/lib/src/commands/daemon.dart [hunk 3]
```diff
@@ -393,16 +401,19 @@ class AppDomain extends Domain {
       _sendAppEvent(app, 'started');
     });
 
-    await app._runInZone(this, () async {
+    await app._runInZone<Null>(this, () async {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/commands/daemon.dart]: `_runInZone<void>`?

File: packages/flutter_tools/lib/src/dart/analysis.dart [hunk 4]
```diff
@@ -136,7 +137,9 @@ class AnalysisServer {
   void _handleAnalysisIssues(Map<String, dynamic> issueInfo) {
     // {"event":"analysis.errors","params":{"file":"/Users/.../lib/main.dart","errors":[]}}
     final String file = issueInfo['file'];
-    final List<AnalysisError> errors = issueInfo['errors']
+    final List<dynamic> errorsList = issueInfo['errors'];
+    final List<AnalysisError> errors = errorsList
+        .map(castStringKeyedMap)
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/dart/analysis.dart]: The two `map` calls should have type parameters `Map<String, dynamic>` and `AnalysisError`, respectively.

File: packages/flutter_tools/lib/src/flutter_manifest.dart [hunk 5]
```diff
@@ -33,13 +34,25 @@ class FlutterManifest {
...
+    } else {
+      pubspec._descriptor = <String, dynamic>{};
+    }
+
+    final Map<dynamic, dynamic> flutterMap = pubspec._descriptor['flutter'];
+    if (flutterMap != null) {
+      pubspec._flutterDescriptor = flutterMap.cast<String, dynamic>();
+    } else {
+      pubspec._flutterDescriptor = <String, dynamic>{};
+    }
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/flutter_manifest.dart]: Maybe extract this pattern to a facility like `castStringKeyedMap`. We could let that method return an empty map on `null` input (but then we might also want to rename it `toStringKeyedMap`).

File: packages/flutter_tools/lib/src/flutter_manifest.dart [hunk 6]
```diff
@@ -105,11 +118,18 @@ class FlutterManifest {
   bool get isModule => moduleDescriptor != null;
 
   List<Map<String, dynamic>> get fontsDescriptor {
-   return _flutterDescriptor['fonts'] ?? const <Map<String, dynamic>>[];
+    final List<dynamic> fontList = _flutterDescriptor['fonts'];
+    return fontList == null
+        ? const <Map<String, dynamic>>[]
+        : fontList.map(castStringKeyedMap).toList();
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/flutter_manifest.dart]: Type parameter on `map`...

This gets annoying. Perhaps we should have
```dart
Map<String, dynamic> toStringKeyedMap(dynamic json) {
  if (json == null) {
    return <String, dynamic>{};
  } else {
    final Map<dynamic, dynamic> map = json;
    return map.cast<String, dynamic>();
  }
}

List<Map<String, dynamic>> toStringKeyedMaps(dynamic json) {
  if (json == null) {
    return <Map<String, dynamic>>[];
  } else {
    final List<dynamic> list = json;
    return list.map<Map<String, dynamic>>(toStringKeyedMap).toList();
  }
}
```
Alternative naming: `toJsonObject`/`toJsonArray`.

File: packages/flutter_tools/lib/src/flutter_manifest.dart [hunk 7]
```diff
@@ -105,11 +118,18 @@ class FlutterManifest {
...
   }
 
   List<Uri> get assets {
-    return _flutterDescriptor['assets']?.map(Uri.encodeFull)?.map(Uri.parse)?.toList() ?? const <Uri>[];
+    final List<dynamic> assets = _flutterDescriptor['assets'];
+    if (assets == null) {
+      return const <Uri>[];
+    }
+    return assets.cast<String>().map(Uri.encodeFull)?.map(Uri.parse)?.toList();
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/flutter_manifest.dart]: Type parameter on `map`, twice.

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 8]
```diff
@@ -95,19 +96,14 @@ class SimControl {
 
     for (String deviceCategory in devicesSection.keys) {
       final List<dynamic> devicesData = devicesSection[deviceCategory];
-      for (Map<String, dynamic> data in devicesData.map(_castStringKeyedMap)) {
+      for (Map<String, dynamic> data in devicesData.map(castStringKeyedMap)) {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: Type parameter on `map`.

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 9]
```diff
@@ -284,7 +288,7 @@ class VMService {
     Map<String, dynamic> params,
   ) {
     return Future.any(<Future<Map<String, dynamic>>>[
-      _peer.sendRequest(method, params),
+      _peer.sendRequest(method, params).then(castStringKeyedMap),
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/vmservice.dart]: Type parameter on `then`.

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 10]
```diff
@@ -256,7 +257,10 @@ class VMService {
 
   /// Whether our connection to the VM service has been closed;
   bool get isClosed => _peer.isClosed;
-  Future<Null> get done => _peer.done;
+
+  Future<Null> get done async {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/vmservice.dart]: `Future<void>`?

**Review by mravn-google — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/flutter_manifest.dart (hunk 6 above)
Comment (mravn-google) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: The empty map and list could be `const`, but then the semantics gets complicated: `toStringKeyedMap(null)` is immutable, whereas `toStringKeyedMap({})` is not. If we do use `const` we should return immutable objects in all cases.

File: packages/flutter_tools/lib/src/commands/daemon.dart (hunk 2 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/commands/daemon.dart]: I think we want `dynamic` here, to match what we get from `_handlers[command](args)`.

File: packages/flutter_tools/lib/src/commands/daemon.dart (hunk 3 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/commands/daemon.dart]: Hmm, at runtime, this seemed to want a `Null`.

File: packages/flutter_tools/lib/src/flutter_manifest.dart (hunk 6 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: Let me look into creating a `toStringKeyedMaps()` method.

File: packages/flutter_tools/lib/src/dart/analysis.dart (hunk 4 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/dart/analysis.dart]: added

File: packages/flutter_tools/lib/src/asset.dart (hunk 1 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/asset.dart]: done

File: packages/flutter_tools/lib/src/flutter_manifest.dart (hunk 7 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: done

File: packages/flutter_tools/lib/src/ios/simulators.dart (hunk 8 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 9 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/vmservice.dart]: added

File: packages/flutter_tools/lib/src/flutter_manifest.dart (hunk 6 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: @mravn-google, I may save that for a follow-up PR (`toStringKeyedMaps`).

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 10 above)
Comment (devoncarew) [reply packages/flutter_tools/lib/src/vmservice.dart]: I really don't know if we're returning `null` or `void` here.

**Review by aam — APPROVED** [review]

**Pull request merged**